rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,1625 @@
+
+The exact design of the tests is an implementation detail, so it's not specified
+in this RFC, but the suggested approach, as a starting point, could be to deploy
+Vector using our documented installation methods, then run some log-generating
+workload and then run assertions on the collected logs.
+
+The things we'd generally want to ensure work properly include (but are not
+limited to):
+
+- basic log collection and parsing
+- log message filtering (both by file paths and by metadata)
+- log events enhancement with metadata
+- partial log events merging
+
+We want the assertions and tests to be cluster-agnostic, so that they work with
+any supplied kubectl config.
+
+#### Existing k8s tests
+
+We already have k8s integration tests implemented in Rust in the Vector repo.
+Currently, they're being run as part of the `cd tests; make tests`. They
+assert that Vector code works properly by deploying Vector plus some test
+log producers and asserting that Vector produced the expected output. This is
+very elegant solution.
+However, these tests are really more like unit tests - in a sense that they
+completely ignore the YAMLs and Helm Charts and
+use their own test configs. While they do a good job in what they're built for -
+we probably shouldn't really consider them integration tests in a broad sense.
+
+It was discussed that we'd want to reuse them as our integration tests, however,
+for the reasons above I don't think it's a good idea. At least as they're now.
+We can decouple the deployment of Vector from the deployment of test containers
+and assertions - then we use just the second half with Vector deployed via YAMLs
+and/or Helm Charts. For now, we should probably leave them as is, maintain them,
+but hold the adoption as integration tests.
+
+### Other data gathering
+
+> This section is on gathering data other than container logs.
+
+While our main focus for the integration is collecting log data from the `Pod`s,
+there are other possibilities to gain observability in the Kubernetes
+environment.
+
+#### Exposing Kubernetes [`Event`s][k8s_api_event] as Vector events
+
+It is possible to subscribe to Kubernetes [`Event`s][k8s_api_event], similarly
+to how this command works:
+
+```shell
+kubectl get events --all-namespaces --watch
+```
+
+Implementing this in Vector would allow capturing the Kubernetes
+[`Event`s][k8s_api_event] and processing them as Vector events.
+
+This feature might be very useful for anyone that wants to see what's going on
+in their cluster.
+
+Note that this feature would require deploying Vector in a differently: instead
+of running Vector on every node, here we need only once Vector instance running
+per cluster. If run on every node, it'd be unnecessarily capturing each event
+multiple times.
+
+So, to implement this, we'd need to add a special source that captures events
+from Kubernetes API, and provide a new workload configuration based on
+[`Deployment`][k8s_api_deployment].
```

Comment:
I agree, I like this strategy. And this should come after we do everything else.